libs/wingc/src/lsp/rename_visitor.rs

```diff
@@ -0,0 +1,167 @@
+use std::sync::Mutex;
+
+use lsp_types::{Position, Range, TextEdit};
+
+use crate::diagnostic::WingLocation;
+use crate::parser::RESERVED_WORDS;
+use crate::type_check::symbol_env::LookupResult;
+use crate::type_check::{SymbolKind, Types};
+use crate::visit::{visit_scope, Visit};
+use crate::visit_context::{VisitContext, VisitorWithContext};
+use crate::{ast::*, visit_context};
+pub struct LinkedSymbol<'a> {
+	symbol: Symbol,
+	references: Mutex<Vec<&'a Symbol>>,
+}
+pub struct RenameVisitor<'a> {
+	types: &'a Types,
+	linked_symbols: Vec<LinkedSymbol<'a>>,
+	ctx: VisitContext,
+}
+
+impl<'a> RenameVisitor<'a> {
+	pub fn new(types: &'a Types) -> Self {
+		Self {
+			types,
+			linked_symbols: vec![],
+			ctx: VisitContext::new(),
+		}
+	}
+
+	fn is_symbol_linked(&mut self, symbol: &'a Symbol) -> bool {
+		self
+			.linked_symbols
+			.iter()
+			.any(|s: &LinkedSymbol<'a>| symbol.same(&s.symbol))
+	}
+
+	fn add_reference_symbol(&mut self, symbol: &'a Symbol) {
+		// symbols that appear in let/if lef statements will point to a prev declaration of a variable of the same name if exists
+		// this is why we add them in advance during visit_statement
+		// the other condition is for "this" that points to the "new" keyword for some reason
+		if self.is_symbol_linked(symbol) || RESERVED_WORDS.contains(&symbol.name) {
+			return;
+		}
+		if let Some(env) = self.ctx.current_env() {
+			match env.lookup_ext(symbol, None) {
+				LookupResult::Found(symbol_kind, lookup_info) | LookupResult::NotPublic(symbol_kind, lookup_info) => {
+					// TODO: remove to support rename-refactor of namespaces - after adjusting the edit
+					if matches!(symbol_kind, &SymbolKind::Namespace(_)) {
+						// as their rename include " as ___"
+						return;
+					}
```

Comment:
We need to handle cases where the span points to an "invisible" span. This is mostly relevant for globals since they aren't necessarily defined in the current file, so you should not be able to rename them. Probably worth adding a test to make sure you can't rename a global.

```suggestion
					if matches!(symbol_kind, &SymbolKind::Namespace(_)) {
						// as their rename include " as ___"
						return;
					}
					if lookup_info.span.is_default() {
					  // This symbol is either generated or otherwise can't be renamed (like a global)
						return;
					}
```

Reply:
With this you probably don't need to check the `RESERVED_WORDS` anymore

Reply:
Like https://github.com/winglang/wing/pull/6461#discussion_r1598869383 I tested this locally but accidentally pushed the commit. Feel free to change/add stuff to my changes!

Reply:
I tested it again and it seems great, thanks for pushing the commit!